### ISSUE TITLE
OpenAIVae implementation

### ISSUE THREAD

**Issue by CDitzel** [lucidrains/DALLE-pytorch#74]
do I see it correctly that the code fragments provided by OpenAI and the way you binded it in the vae.py file means that there is no actual codebook in the form of an explicit nn.Paramter or nn.Embeddings but the very first layer of the decoder serves as the vocabulary?

```
(decoder): Decoder(
    (blocks): Sequential(
      (input): Conv2d(n_in=8192, n_out=128, kw=1, use_float16=False, device=device(type='cpu'), requires_grad=False)
```

this would explain why I couldnt finy any oO

**Comment by lucidrains** [comment]
@CDitzel that's exactly my understanding Carsten :)

**Comment by CDitzel** [comment]
all right. So then there is not Gumbel Softmax or reparameterization trick at work at all. 

Also no explicit KL loss to consider as there are no constraints placed on the "codebook" to begin with.

That would explain why the first author of the paper struggled to explain the loss formula he used in his own paper when Andrej and Justin asked him about it in clubhouse oO

**Comment by CDitzel** [comment]
of course Phil. Happy to be of service. Its [timestamped](https://youtu.be/PtdpWC7Sr98?t=2544) to the correct position.

Mh you are right, they do mention Gumbel softmax in the paper but I am wondering how when and where.

I tried to finetune the given OpenAI models after downloading them with your skript but didnt manage to set "requires_grad" to True. Do you happen to know if it is possible to prohibit that and if so if OpenAI did that to their model?

**Comment by lucidrains** [comment]
the gumbel softmax would be used on the encoder output, just before it is sent back to the decoder (where-as the hard version uses a one-hot)

nope, you can't prohibit that! once you have the weights, the sky is the limit

**Comment by CDitzel** [comment]
but its hard trying to finetune the provided encoder and decoder when I still neither understoof completely their data regime mapping procedure nor the specifics of their loss function. Trying simple L1 loss and backpropping through their networks yield NaN pretty soon...

**Comment by CDitzel** [comment]
I found that the Gumbel Softmax significantly impairs the reconstruction quality. Leaving it out and just doing the tensor contraction defaults to OpenAIs provided implementation. However, with their code I cant seem to get satisfying reconstructions with tokens via the one_hot encoding scheme. Crazy...

**Comment by fractaldna22** [comment]
@CDitzel  > "I found that the Gumbel Softmax significantly impairs the reconstruction quality. Leaving it out and just doing the tensor contraction defaults to OpenAIs provided implementation. However, with their code I cant seem to get satisfying reconstructions with tokens via the one_hot encoding scheme. Crazy..." \

https://colab.research.google.com/drive/1oA1fZP7N1uPBxwbGIvOEXbTsq2ORa9vb?usp=sharing#scrollTo=O78RfTZfh7ji lol 

I find that[ this notebook](https://colab.research.google.com/drive/1oA1fZP7N1uPBxwbGIvOEXbTsq2ORa9vb?usp=sharing#scrollTo=O78RfTZfh7ji) solves all the problems and makes it much easier to get a grasp on it.  I feel that it's a waste of time trying to train a not very deep "discrete" VAE which is like 90 mb - when it took tech giants like 3 weeks on a million tpus to get a pretrained model which they didn't even release. The VAE they release is practically just an encoder and a decoder - there are plenty. 

The REAL magic of  Dall_E is not the VAE - but CLIP itself. Behind CLIP is the VIsual Transformer - VIT. I don't see ViT mentioned in this notebook at all - it focuses exclusively on the vae, which is literally just a middle man between CLIP and the model. Clip is only mentioned in passing and nothing is said about what model you should use or what the architecture for that is.  Just my feel.   I spent days trying to get this to do anything and i feel like those hours were in VAEn.  Is this all just to get a graph? Wheres the pics lol. Where's the hedgehog violin?

 [Dall-E = A normalized latent space to start, text, tokenizer, mapper +  ** ViT **](https://colab.research.google.com/drive/1oA1fZP7N1uPBxwbGIvOEXbTsq2ORa9vb?usp=sharing#scrollTo=O78RfTZfh7ji)

Correct me if I'm wrong, but from my very amateur understanding, but having already generated hundreds of dall-e-like images in practice, this is my two cents:

The central player in this implementation is Clip.load('ViT-B/32', jit=True)' -- the VisualTransformer in this, ViT-B-32.pt  343 mb which is far larger than the 'discrete' 'tinyVAE' and far more robust,  the VIT pretrained model is the real performer. It's trained on thousands of imagenet images - this is where the magic happens.  As far as encoding and decoding of text to tokens, and pixels to tensor numbers -- any language model will do. You can use simple.tokenizer. you can use bert, or gpt2. You can use VAE. It's a fairly linear translation from a very small dictionary that frankly could be bigger afaic. Convert text to tokens, and then send those tokens to a visual transformer model so it knows what categories it should force-hallucinate into existence from the latent space --  reward the mapper for increasing the similarity to the categories, and penalize the mapper if the similarity falls --until the mapper gets its p*xels together. .  

The latest version of this notebook even decided to drop the dall_e  encoder completely and it didn't even change the result except that it uses less memory. The encoder is super inefficient in compute. Perceptor(clip.model(ViT-size-int) is just as good at it, knows better than it, and is pretrained on a lot of images you'll see in the Multi-Modal Neurons distil.pub ;   VAE encoder seems redundant  when you already have simple.tokenizer in the clip model doing the same thing. What you don't have is a visual perceiver that tells the pixel mapper how good it's doing until it learns that moving the latent space pixels in xyz directions achieves less loss and more similarity to the text_input, than moving the pixels in incorrect directions.  

Using the pretrained ViT, with the proper setup of the temp / tau, lr (.1 works fine), ZERO_GRAD,  ncols, mean, std, and .clipping min and max  --  it achieves similar Image examples of the type:  "Hedgehogs made out of legos" and whatever you want.  This is Dall_E : CLIP using Google's VIsual Transformer model or a similar model trained on a HUGE dataset + a simple tokenizer and decoder. 

The way forward is to train a Visual Transformer model using clip - with or without the involvement of miniVAE (I would think a very deep vae would be better), on as much images as possible. CLIP handles the labeling of them, that's what its for!  Clip CAN see why kids love cinnamon toast crunch.   

The ViT itself has a built in encoder and decoder, attention, etc and maps the pixels.   This Attention is ALL you need.  if you have a pretrained ViT, and the ViT knows what it's looking at, and CLIP knows how concepts relate, and relay the right amount of reward and penalty to get the mapper correct itself --  then you get the image you want to see.  If you only have a VAE that only knows how to encode and decode from text to token and token to array -- its still not good or much If it doesn't even know what it's looking at or even see the results. 

So far based on this paper - https://www.kaggle.com/abhinand05/vision-transformer-vit-tutorial-baseline# -  it looks like ViT-H-14 and ViT-L-16 are the best , achieving up to 99.74% accuracy between image and text predictions or vice versa. 
 
The notebook i linked possibly builds the 'vae' architecture into the "perceptor' state_dict but i'm not sure on that. Take a look at it if you haven't seen it.  I would really love to try loading ViT-L-16 or ViT-H-41 into perceptor / clip but so far I can't figure out how to configure the .PT with the correct keys like "version" so it rejects my attempts at loading in custom models.

**Comment by fractaldna22** [comment]
ViT > VaE lol

**Comment by alexisrozhkov** [comment]
@CDitzel I'm not an expert on the subject, but my understanding is that DALL-E paper authors didn't use "vanilla" VQ-VAE (that requires explicit codebook that you were looking for). 

Everywhere in the paper they refer to "dVAE", for instance:
> We train a discrete variational autoen-
> coder (dVAE) to compress each 256×256 RGB image
> into a 32 × 32 grid of image tokens, each element of
> which can assume 8192 possible values.

It makes me think that they have 32x32x8192 logits output that is sampled/argmax'ed to obtain 32x32x1 token tensor - this works with the code and weights for VAE that they've shared publicly.

> Also no explicit KL loss to consider as there are no constraints placed on the "codebook" to begin with.
I don't see any issue in applying KL loss to the logits I've mentioned above.

> So then there is not Gumbel Softmax or reparameterization trick at work at all.
How exactly did you arrive to this conclusion?

**Comment by fractaldna22** [comment]
![image](https://user-images.githubusercontent.com/37323518/113177250-4e474d00-921b-11eb-935d-23d4eee0a394.png)
vqvae > tiny-vae(big memory) aka dall-e the model that never was

**Comment by fractaldna22** [comment]
+ it does animations "A girl walking on the sidewalk in slow motion"  produces iterations of a girl actually walking in slow motion. Dall-E has no working model or else they'd let you choose your own inputs.  

The real powerhouse behind Dall-E is clip. Without clip its nothing but a tokenizer. Clip is what guides the mapping and unmapping of pixels.  OpenAI has both -- they arent going to simply not use clip for literally everything that has to do with image+texts lol. 

The vae that is given here is one of the worst. Its a generic dollar store vae they give you just so they can say the claim is scientific because they provided you with a crayon, without releasing the actual model to the Open internet its pseudo-science

**Comment by CDitzel** [comment]
I have to agree with you on this one. OpenAIs policy of reproducible research and being honest/open about their work has been questionable, to say the least. I am going to post the vae-gumbel softmax snippet here soon as a reference and so that we can talk about it. First have to get coffee and breakfast though xD

**Comment by CDitzel** [comment]
all right, so this is what I have come up with so far. It closely resembles Lucids implementations but parameterizes the gumbel softmax with the distance of the encoder output (logits) to the codebook vectors (described in [this paper](http://bayesiandeeplearning.org/2017/papers/54.pdf)) and akin to VQ-VAEs, but in contrast to Lucids implementation which uses the logits directly as input to the Gumbel. Phils (and Karpathys implementation) never worked for me when I rightfully included the KL loss i.e. a kl loss > 0. With this implementation the KL loss can be included as it should with a uniform prior. However, the results on a larger data set are still underwhelming and not really satisfying in terms of reconstruction quality. Maybe someone can take a look at it and assess the correctness of this implementation?

```
class SoftDiscretizer(nn.Module):
    def __init__(
        self,
        nTokens,
        dTokens,
        temperature,
        kl_weight,
        **kwargs,
    ):
        super().__init__()
        self.nTokens = nTokens
        self.dTokens = dTokens
        self.kl_weight = kl_weight

        self.embedding = nn.Embedding(nTokens, dTokens)

    def forward(self, z):
        B, C, H, W = z.size()
        N, D = self.embedding.weight.shape

        z_flat = rearrange(z, "b c h w -> (b h w) c")
        distances = (
            torch.sum(self.embedding.weight ** 2, dim=1)
            + torch.sum(z_flat ** 2, dim=1, keepdim=True)
            - 2 * torch.matmul(z_flat, self.embedding.weight.t())
        )
        distances = rearrange(distances, "(b h w) n -> b h w n", h=H, w=W)

        # minus so that closer codebook vectors have higher probability?
        samples = F.gumbel_softmax(-distances, temperature, hard=hard, dim=-1)

        if not self.training:
            tokens = samples.argmax(dim=-1)
            return tokens.flatten(start_dim=1)

        z_q = einsum("b h w n, n d -> b d h w", samples, self.embedding.weight)

        # KL loss
        logits = F.log_softmax(-distances, dim=-1)
        probs = torch.exp(logits)  # supposed to be numerically more stable than softmax alone
        neg_entropy = torch.sum(probs * (logits + math.log(self.nTokens)), dim=(1, 2, 3))
        kl_loss = self.kl_weight * torch.mean(neg_entropy)

        return z_q, kl_loss
```

**Comment by afiaka87** [comment]
> ![image](https://user-images.githubusercontent.com/37323518/113177250-4e474d00-921b-11eb-935d-23d4eee0a394.png)
> vqvae > tiny-vae(big memory) aka dall-e the model that never was

Do you have a method for implementing the "image mask" feature? I've done extensive scraping of that blog post in particular (check the discussions tab for a scrape of all 1.1 million image text pairs) and they aren't always clear about when they're using that feature.

It's possible they're using a mask containing white pixels at the sides and a transparent square cropped in the center.

**Comment by fractaldna22** [comment]
Yes totally. And I don't see the words "white background" in the prompt

**Comment by sidml** [comment]
@CDitzel 
You may want to check out [this issue](https://github.com/karpathy/deep-vector-quantization/issues/7).
The gumbel softmax function expects logits as inputs, so passing distances mayn't be appropriate. The scale of distances will be quite different from the scale of gumbel distribution (which is between 0 and 1).

**Comment by enhuiz** [comment]
> @CDitzel
> You may want to check out [this issue](https://github.com/karpathy/deep-vector-quantization/issues/7).
> The gumbel softmax function expects logits as inputs, so passing distances mayn't be appropriate. The scale of distances will be quite different from the scale of gumbel distribution (which is between 0 and 1).

But logits are not between 0 and 1 either, could you please elaborate on "quite different from the scale of gumbel distribution"?

**Comment by sidml** [comment]
@enhuiz 
The standard Gumbel distribution i.e. G(0, 1) looks like this
![image](https://user-images.githubusercontent.com/51043033/117124412-dd232880-adca-11eb-8b90-985aaf534ad5.png)

When we call F.gumbel_softmax in pytorch, it adds the provided logits with G(0, 1) distribution.
If the logits are at some wierd scale, then it mayn't make sense to add it with the sample
form gumbel distribution.
![image](https://user-images.githubusercontent.com/51043033/117124029-61c17700-adca-11eb-9ea0-340bbc2e1a21.png)

More details are [here](https://arxiv.org/abs/1611.01144)

**Comment by CDitzel** [comment]
but just passing the encoder outputs/logits doesnt make any guarantees on the value ranges either, does it?

**Comment by enhuiz** [comment]
@sidml Thanks for the reply. I agree that logits should have a similar range with g_i (i.e., roughly from -2 to 6), but since the latent space is learnable, I guess the scale of the distance can be learned to suits the scale of g_i.

**Comment by CDitzel** [comment]
have a look at this notebook https://github.com/shaabhishek/gumbel-softmax-pytorch/blob/master/Gumbel-softmax%20visualization.ipynb

here they normalize the logits before passing them to Gumbel-Softmax. Apparently that is not necessary for the built-in Pytorch implementation. 

On another note, can someone explain to me the top row of Figure 1 in the [original paper](https://arxiv.org/pdf/1611.01144.pdf)?

It says expectation but how is that calculated or even depending on the temperature parameter?

**Comment by sidml** [comment]
@CDitzel 
In [pytorch implementation](https://github.com/pytorch/pytorch/blob/4cb534f92ef6f5b2ec99109b0329f93a859ae831/torch/csrc/api/include/torch/nn/functional/activation.h#L177), they seem to be directly adding the logits to the sample from gumbel distribution.

I believe they divide the logits by temperature before sampling from categorical distribution in Figure 1 of the paper.

**Comment by LordHui** [comment]
> all right, so this is what I have come up with so far. It closely resembles Lucids implementations but parameterizes the gumbel softmax with the distance of the encoder output (logits) to the codebook vectors (described in [this paper](http://bayesiandeeplearning.org/2017/papers/54.pdf)) and akin to VQ-VAEs, but in contrast to Lucids implementation which uses the logits directly as input to the Gumbel. Phils (and Karpathys implementation) never worked for me when I rightfully included the KL loss i.e. a kl loss > 0. With this implementation the KL loss can be included as it should with a uniform prior. However, the results on a larger data set are still underwhelming and not really satisfying in terms of reconstruction quality. Maybe someone can take a look at it and assess the correctness of this implementation?
> 
> ```
> class SoftDiscretizer(nn.Module):
>     def __init__(
>         self,
>         nTokens,
>         dTokens,
>         temperature,
>         kl_weight,
>         **kwargs,
>     ):
>         super().__init__()
>         self.nTokens = nTokens
>         self.dTokens = dTokens
>         self.kl_weight = kl_weight
> 
>         self.embedding = nn.Embedding(nTokens, dTokens)
> 
>     def forward(self, z):
>         B, C, H, W = z.size()
>         N, D = self.embedding.weight.shape
> 
>         z_flat = rearrange(z, "b c h w -> (b h w) c")
>         distances = (
>             torch.sum(self.embedding.weight ** 2, dim=1)
>             + torch.sum(z_flat ** 2, dim=1, keepdim=True)
>             - 2 * torch.matmul(z_flat, self.embedding.weight.t())
>         )
>         distances = rearrange(distances, "(b h w) n -> b h w n", h=H, w=W)
> 
>         # minus so that closer codebook vectors have higher probability?
>         samples = F.gumbel_softmax(-distances, temperature=0.5, hard=False, dim=-1)
> 
>         if not self.training:
>             tokens = samples.argmax(dim=-1)
>             return tokens.flatten(start_dim=1)
> 
>         z_q = einsum("b h w n, n d -> b d h w", samples, self.embedding.weight)
> 
>         # KL loss
>         logits = F.log_softmax(-distances, dim=-1)
>         probs = torch.exp(logits)  # supposed to be numerically more stable than softmax alone
>         neg_entropy = torch.sum(probs * (logits + math.log(self.nTokens)), dim=(1, 2, 3))
>         kl_loss = self.kl_weight * torch.mean(neg_entropy)
> 
>         return z_q, kl_loss
> ```

For test procedure, may it need to make some changes here?
```
        if not self.training:
            # tokens = samples.argmax(dim=-1)
            tokens = distance.argmin(dim=-1)
            return tokens.flatten(start_dim=1)
```

**Comment by thuangb** [comment]
I also tried all existing gumbel quantizer but they all perform worse than normal quantizer. But I've recently notice from OpenAI's code for DALL-E, the encoder outputs the probability and take that directly as the input for the decoder without using any codebook embedding. Which leads me to following dummy gumbel quantizer and I notice that with some hyper-parameter tweaking, it even works better than or on par with previous gumbel quantizer implementations:

``` 
class DummyDiscretizer
     def __init__(
        self,
        z_dim,
        nTokens,
        temperature,
        **kwargs,
       ):

       super().__init__()
       self.z_dim = z_dim

       self.nTokens = nTokens
       self.temp = temperature
       
       self.proj_in =  nn.Conv2d(z_dim, nTokens, kernel_size=1)
       self.proj_out =  nn.Conv2d(nTokens, z_dim, kernel_size=1)

    def forward(self, z):
        B, C, H, W = z.size()

        logits = self.proj_in(z).exp()
        probs = F.gumbel_softmax(logits, temperature=self.temp, hard=False, dim=-1)

        if not self.training:
            tokens = probs .argmax(dim=-1)
            return tokens.flatten(start_dim=1)

        z_q = self.proj_out(probs) 

        # Dummy Loss, can improve more with OpenAI myterious kl loss =)))
        kl_loss = torch.Tensor(0)

        return z_q, kl_loss
```

I think this means that something is wrong with the KL loss because it is almost useless
